src/main/java/life/qbic/data/processing/registration/ProcessRegistrationRequest.java

```diff
@@ -40,53 +52,134 @@ public class ProcessRegistrationRequest extends Thread {
-        // TODO move back to user folder
-      } catch (IOException e) {
-        log.error("Error while processing registration request", e);
-        // TODO move back to user folder
-      } finally {
-        active.set(false);
-        log.info("Processing completed: {}", request);
+  private static void cleanup(Path workingTargetDir) throws IOException {
+    try (var content = Files.walk(workingTargetDir)) {
+      content.map(Path::toFile).sorted(Comparator.reverseOrder()).forEach(File::delete);
+    }
+  }
+
+  private static void processMeasurement(String measurementId,
+      Map<String, List<RegistrationMetadata>> aggregatedFilesByMeasurementId, Path workingTargetDir,
+      Path taskDir) throws IOException {
+    for (RegistrationMetadata metadataEntry : aggregatedFilesByMeasurementId.get(
+        measurementId)) {
+      Files.move(workingTargetDir.resolve(metadataEntry.file()),
+          taskDir.resolve(metadataEntry.file()));
+    }
+  }
+
+  private void moveBackToOrigin(Path target, Path usersHomePath, String reason) {
+    log.info("Moving back to original user directory: {}", usersHomePath);
+    try {
+      Path taskDir = createTaskDirectory();
+      Files.move(target, taskDir.resolve(target.getFileName()));
+      var errorFile = taskDir.resolve("error.txt").toFile();
+      errorFile.createNewFile();
+      Files.writeString(errorFile.toPath(), reason);
+      usersHomePath.resolve(usersErrorDirectory).toFile().mkdir();
+      Files.move(taskDir,
+          usersHomePath.resolve(usersErrorDirectory)
+              .resolve(taskDir.toFile().getName()));
+    } catch (IOException e) {
+      log.error("Cannot move task to user intervention: %s".formatted(
+          usersHomePath.resolve(usersErrorDirectory)), e);
+    }
+  }
+
+  private void validateFileEntries(Collection<RegistrationMetadata> metadata, Path request)
+      throws ValidationException {
+    for (RegistrationMetadata metadataEntry : metadata) {
+      if (!request.resolve(Paths.get(metadataEntry.file())).toFile().exists()) {
+        throw new ValidationException(
+            "Unknown file reference in metadata: %s".formatted(metadataEntry.file()),
+            ErrorCode.FILE_NOT_FOUND);
+      }
+    }
+    // To save the user from any trouble, let's also check if all dataset files are described by a metadata entry
+    var filesInMetadata = metadata.stream().map(RegistrationMetadata::file).toList();
+    var physicalFiles = request.toFile().listFiles();
+    for (File file : physicalFiles) {
+      // we ignore hidden files
+      if (file.isHidden()) {
+        continue;
+      }
+      if (file.getName().endsWith(metadataFileName)) {
+        continue;
+      }
+      if (!filesInMetadata.contains(file.getName())) {
+        throw new ValidationException(
+            "Found more files than described in the metadata file: %s".formatted(file.getName()),
+            ErrorCode.MISSING_FILE_ENTRY);
```

Comment:
If i understood it correctly, this means that every file has to be provided with the correct filename specified in the metadata file. Is this the intended behaviour?

Reply:
Yes, otherwise files not mention in the metadata file will not get processed and deleted, which we must not let happen. It is just a fail safe for any user error